### PR TITLE
Add tests for duplicate primary keys inserts

### PR DESCRIPTION
**Pull request by Aerion**
* Add a simple test for adding two different products with the same ID
  Note: this is disabled for SqlServer as it would need
  IDENTITY_INSERT enabled.
* Add a test with a composite primary key for adding two different
  items with the same primary keys.
  This test works for all providers.
  Note: this test introduces a dummy entity, CompisiteKeyItem,
  referencing the IDs of Product and ProductSale

These two tests require to have two DbContext in the test. Thus, a new
method returning a new DbContext has been created and the constructors
modified to fit those needs.

### CODE REVIEW

**Comment by Aerion** [comment]
FYI, this is the result of the test after on a branch with both #14 and #15: https://ci.appveyor.com/project/Aerion/entityframework-exceptions/builds/32101410/tests
You can see that in the current state, some tests aren't caught in Sqlite and that when #14 is merged, they'll pass.

File: EntityFramework.Exceptions.Tests/DatabaseTests.cs [hunk 1]
```diff
@@ -7,10 +7,17 @@ namespace EntityFramework.Exceptions.Tests
...
 
-        protected DatabaseTests(DemoContext context)
+        protected DatabaseTests(DbContextOptions<DemoContext> contextOptions)
         {
-            Context = context;
+            ContextOptions = contextOptions;
+            Context = GetNewContext();
+        }
+
+        protected DemoContext GetNewContext()
```
Comment (Giorgi) [inline EntityFramework.Exceptions.Tests/DatabaseTests.cs:18]: Why is it protected?

File: EntityFramework.Exceptions.Tests/SqlServerTests.cs [hunk 2]
```diff
@@ -8,7 +8,12 @@ namespace EntityFramework.Exceptions.Tests
 {
     public class SqlServerTests : DatabaseTests, IClassFixture<SqlServerDemoContextFixture>, IDisposable
     {
-        public SqlServerTests(SqlServerDemoContextFixture fixture) : base(fixture.Context)
+        public SqlServerTests(SqlServerDemoContextFixture fixture) : base(fixture.ContextOptions)
+        {
+        }
+
+        [Fact(Skip = "Skipping as IDENTITY_INSERT must be set to ON to write IDs directly in db")]
```
Comment (Giorgi) [inline EntityFramework.Exceptions.Tests/SqlServerTests.cs:15]: Without actually testing it how do you now that inserting two items with the same PK results in error 2601 or 2627 ? Did you test it manually?

File: EntityFramework.Exceptions.Sqlite/SqliteExceptionProcessorStateManager.cs [hunk 3]
```diff
@@ -44,7 +44,8 @@ public static DbContextOptionsBuilder UseExceptionProcessor(this DbContextOption
             return self;
         }
 
-        public static DbContextOptionsBuilder<TContext> UseExceptionProcessor<TContext>(this DbContextOptionsBuilder<TContext> self) where TContext : DbContext
+        public static DbContextOptionsBuilder<TContext> UseExceptionProcessor<TContext>(
+            this DbContextOptionsBuilder<TContext> self) where TContext : DbContext
```
Comment (Giorgi) [inline EntityFramework.Exceptions.Sqlite/SqliteExceptionProcessorStateManager.cs:48]: Please don't change formatting

Comment (Aerion) [reply EntityFramework.Exceptions.Sqlite/SqliteExceptionProcessorStateManager.cs]: Sure, my bad!

File: EntityFramework.Exceptions.Tests/DatabaseTests.cs (hunk 1 above)
Comment (Aerion) [reply EntityFramework.Exceptions.Tests/DatabaseTests.cs]: There is indeed no need, I changed it to private now. Thanks for pointing that out!

File: EntityFramework.Exceptions.Tests/SqlServerTests.cs (hunk 2 above)
Comment (Aerion) [reply EntityFramework.Exceptions.Tests/SqlServerTests.cs:15]: I'm not sure to understand your question, let me know if I failed to answer it. I suppose in this message that we're only talking about SqlServer.

Msdn tells us [that](https://docs.microsoft.com/en-us/sql/relational-databases/errors-events/database-engine-events-and-errors?view=sql-server-ver15#errors-2000-to-2999) about errors:
* 2601 `Cannot insert duplicate key row in object '%.*ls' with unique index '%.*ls'.`
* 2627 `Violation of %ls constraint '%.*ls'. Cannot insert duplicate key in object '%.*ls'.`

It's still a bit cryptic for me, so I went on another website:
* [2601](http://www.sql-server-helper.com/error-messages/msg-2601.aspx)
* [2627](http://www.sql-server-helper.com/error-messages/msg-2627.aspx)

From these two links, my understanding is that 2601 will trigger when we specify a unique constraint and that it's not respected. While 2627 will trigger for PK constraint not respected.

So I'd guess that inserting two items with the same PK always result in error 2627.
But I didn't manually test the exception for these two, I only launched the automated tests.

Comment (Giorgi) [reply EntityFramework.Exceptions.Tests/SqlServerTests.cs:15]: Yes, that's what I wanted to know. I also think that 2627 is thrown when the PK is violated but it would be nice if this was tested at least manually. By the way what if you use `Context.Database.ExecuteSqlCommand` to set `IDENTITY_INSERT ` on and off before and after the test ?

Comment (Aerion) [reply EntityFramework.Exceptions.Tests/SqlServerTests.cs]: Sure, I did it in 37bad0544f that I just pushed, but it required a bit of work.
I put it in a separate commit so that if you feel that it's too complex or not needed, we can throw it away easily from this PR. But if you're okay with this commit, I'll squash it.

Comment (Giorgi) [reply EntityFramework.Exceptions.Tests/SqlServerTests.cs]: Yes, it looks more complicated to me as I think it needs to be.

**Comment by Giorgi** [comment]
Implemented in 5dc8d5ece1c120d546a171056606a31f2fe91903
